src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,170 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.NoSuchElementException;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            addFirstNode(value);
+        } else {
+            addLastNode(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        Node<T> newNode = new Node<>(null, value, null);
+        if (isEmpty()) {
+            addFirstNode(value);
+        } else if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size) {
+            addLastNode(value);
+        } else if (index == size - 1) {
+            newNode.next = getNode(index);
+            newNode.prev = getNode(index).prev;
+            getNode(index).prev.next = newNode;
+            getNode(index).next = null;
+        } else {
+            newNode.next = getNode(index);
+            newNode.prev = getNode(index).prev;
+            getNode(index).prev.next = newNode;
+            getNode(index).next.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size);
+        T replacedValue = getNode(index).value;
+        getNode(index).value = value;
+        return replacedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        T removedValue = get(index);
+        unlink(getNode(index));
+        size--;
+        return removedValue;
     }
 
     @Override
-    public boolean remove(T object) {
+   public boolean remove(T object) {
+        Node<T> currentNode = head;
+        do {
+            if (currentNode.value == null && object == null
+                    || currentNode.value != null && currentNode.value.equals(object)) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+
+        } while (currentNode.next != null);
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index, int size) {
```

Comment:
```suggestion
    private void checkIndex(int index) {
```
Why pass the size here if you already have access to the field?